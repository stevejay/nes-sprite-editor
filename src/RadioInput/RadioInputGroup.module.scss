.radioGroup {
  margin-bottom: 1rem;

  label {
    margin-bottom: 0.5rem;

    &:last-of-type {
      margin-bottom: 0;
    }
  }
}
