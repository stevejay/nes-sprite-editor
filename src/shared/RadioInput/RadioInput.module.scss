@import "styles/_base.scss";

$input-color: $md-grey-800;
$disabled-color: $md-grey-200;

.input {
  @include screen-reader-only;
}

.input + label {
  position: relative;
  display: inline-block;
  cursor: pointer;
  padding: rem(2) rem(1) rem(2) rem(25);
  // border-radius: rem(2);
}

// .input + label:hover {
//   background-color: $md-grey-200;
// }

.input + label::before,
.input + label::after {
  content: "";
  position: absolute;
  left: rem(10);
  top: 50%;
  transform: translate(-50%, -50%);
  border-radius: rem(9);
}

.input + label::before {
  height: rem(18);
  width: rem(18);
  border: 2px solid $input-color;
}

.input + label::after {
  height: rem(10);
  width: rem(10);
  background-color: transparent;
  transition: background-color 0.3s;
}

input:checked + label::after {
  background-color: $input-color;
}

.input:disabled + label {
  cursor: not-allowed;
}

.input:disabled + label::before {
  border-color: darken($disabled-color, 30%);
  background-color: $disabled-color;
}

.input:disabled + label::after {
  background-color: darken($disabled-color, 10%);
}

.input:focus + label {
  @include focus-ring;
}
