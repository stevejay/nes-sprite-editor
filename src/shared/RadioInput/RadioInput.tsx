import React from "react";
import styles from "./RadioInput.module.scss";
import useId from "../utils/use-id";

type Props = {
  value: string | number;
  children: React.ReactNode;
  checked: boolean;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  groupName?: string;
  disabled?: boolean;
};

const RadioInput = ({
  value,
  children,
  checked,
  onChange,
  groupName,
  disabled = false
}: Props) => {
  const id = useId("radio-input_");
  return (
    <>
      <input
        type="radio"
        id={id}
        name={groupName}
        value={value}
        checked={checked}
        disabled={disabled}
        className={styles.input}
        onChange={onChange}
      />
      <label htmlFor={id}>{children}</label>
    </>
  );
};

export default RadioInput;
