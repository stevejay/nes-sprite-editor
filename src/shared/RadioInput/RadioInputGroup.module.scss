@import "styles/_base.scss";

.label {
  margin-bottom: 0.5rem;
}

.groupContainer {
  margin-bottom: 1rem;
}

.inlineInputsContainer {
  display: inline-block;
  margin-right: rem(8);

  &:last-of-type {
    margin-right: 0;
  }
}

.blockInputsContainer {
  margin-bottom: rem(8);

  &:last-of-type {
    margin-bottom: 0;
  }
}
