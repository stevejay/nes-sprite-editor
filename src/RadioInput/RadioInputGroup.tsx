import React from "react";
import { uniqueId } from "lodash";
import RadioInput from "./RadioInput";
import styles from "./RadioInputGroup.module.scss";

type Props<IdType> = {
  legend: string;
  options: Array<{ id: IdType; label: string }>;
  selectedId: IdType;
  onChange: (id: IdType) => void;
};

const RadioInputGroup: React.FunctionComponent<Props<string>> = ({
  legend,
  options,
  selectedId,
  onChange
}) => {
  const labelId = React.useRef(uniqueId("radio-group-label_"));

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChange(event.target.value);
  };

  return (
    <>
      <p id={labelId.current}>{legend}</p>
      <div
        role="radiogroup"
        className={styles.radioGroup}
        aria-labelledby={labelId.current}
      >
        {options.map(option => (
          <React.Fragment key={option.id}>
            <RadioInput
              value={option.id}
              label={option.label}
              checked={selectedId === option.id}
              onChange={handleChange}
            />
            <br />
          </React.Fragment>
        ))}
      </div>
    </>
  );
};

export default RadioInputGroup;
