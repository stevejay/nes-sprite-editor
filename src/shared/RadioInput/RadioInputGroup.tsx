import React from "react";
import { uniqueId } from "lodash";
import RadioInput from "./RadioInput";
import styles from "./RadioInputGroup.module.scss";

type Option = {
  id: string | number;
  label: string;
};

type Props = {
  legend: string;
  options: Array<Option>;
  selectedId: string | number;
  disabled?: boolean;
  inline?: boolean;
  renderLabel?: (option: Option) => React.ReactNode;
  onChange: (id: string | number) => void;
};

const RadioInputGroup: React.FunctionComponent<Props> = ({
  legend,
  options,
  selectedId,
  disabled,
  inline,
  renderLabel = (option: Option) => option.label,
  onChange
}) => {
  const labelId = React.useRef(uniqueId("radio-group_"));

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChange(event.target.value);
  };

  return (
    <>
      <p id={labelId.current}>{legend}</p>
      <div
        role="radiogroup"
        className={styles.groupContainer}
        aria-labelledby={labelId.current}
      >
        {options.map(option => (
          <div
            key={option.id}
            className={inline ? styles.inlineInput : styles.blockInput}
          >
            <RadioInput
              value={option.id}
              groupName={labelId.current}
              checked={selectedId === option.id}
              disabled={disabled}
              onChange={handleChange}
            >
              {renderLabel(option)}
            </RadioInput>
          </div>
        ))}
      </div>
    </>
  );
};

export default RadioInputGroup;
