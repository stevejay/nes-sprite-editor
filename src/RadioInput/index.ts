import RadioInput from "./RadioInput";
import RadioInputGroup from "./RadioInputGroup";

type RadioInputFn = typeof RadioInput & { Group: typeof RadioInputGroup };

const RadioInputExport = RadioInput as RadioInputFn;
RadioInputExport.Group = RadioInputGroup;

export default RadioInputExport;
