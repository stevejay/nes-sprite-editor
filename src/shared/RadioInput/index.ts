import RadioInput from "./RadioInput";
import RadioInputGroup from "./RadioInputGroup";

type RadioInputType = typeof RadioInput & { Group: typeof RadioInputGroup };

const RadioInputExport = RadioInput as RadioInputType;
RadioInputExport.Group = RadioInputGroup;

export default RadioInputExport;
