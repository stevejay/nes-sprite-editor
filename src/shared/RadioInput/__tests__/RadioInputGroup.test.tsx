import React from "react";
import { render } from "@testing-library/react";
import { noop } from "lodash";
import userEvent from "@testing-library/user-event";
import RadioInput from "..";

const options = [
  { id: 1, label: "First option" },
  { id: 2, label: "Second option" }
];

test("displays correctly", async () => {
  const { container } = render(
    <RadioInput.Group
      legend="The Radio Group Legend:"
      options={options}
      selectedId={2}
      onChange={noop}
    />
  );
  expect(container).toMatchSnapshot();
});

test("has correct a11y markup", async () => {
  const { getByRole, getByLabelText } = render(
    <RadioInput.Group
      legend="The Radio Group Legend:"
      options={options}
      selectedId={2}
      onChange={noop}
    />
  );

  const radioGroup = getByRole("radiogroup");
  expect(radioGroup).toBeTruthy();

  const radioGroupByLegend = getByLabelText("The Radio Group Legend:");
  expect(radioGroupByLegend).toBeTruthy();
  expect(radioGroupByLegend).toEqual(radioGroup);

  const radioInputOne = getByLabelText("First option");
  expect(radioInputOne).toBeTruthy();

  const radioInputTwo = getByLabelText("Second option");
  expect(radioInputTwo).toBeTruthy();
});

test("change selected input from second to first", async () => {
  const handleChange = jest.fn();
  const { getByLabelText } = render(
    <RadioInput.Group
      legend="The Radio Group Legend:"
      options={options}
      selectedId={2}
      onChange={handleChange}
    />
  );

  userEvent.click(getByLabelText("First option"));
  expect(handleChange).toHaveBeenCalledWith(1);
});
