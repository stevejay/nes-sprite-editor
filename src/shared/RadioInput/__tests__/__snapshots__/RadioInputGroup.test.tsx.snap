// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays correctly 1`] = `
<div>
  <p
    class="label"
    id="radio-group_1"
  >
    The Radio Group Legend:
  </p>
  <div
    aria-labelledby="radio-group_1"
    class="groupContainer"
    role="radiogroup"
  >
    <div
      class="blockInputsContainer"
    >
      <input
        class="input"
        id="radio-input_2"
        name="radio-group_1"
        type="radio"
        value="1"
      />
      <label
        for="radio-input_2"
      >
        First option
      </label>
    </div>
    <div
      class="blockInputsContainer"
    >
      <input
        checked=""
        class="input"
        id="radio-input_3"
        name="radio-group_1"
        type="radio"
        value="2"
      />
      <label
        for="radio-input_3"
      >
        Second option
      </label>
    </div>
  </div>
</div>
`;
